src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.cpp

```diff
@@ -148,6 +149,15 @@ bool PrimitiveFromConservative<OrderedListOfPrimitiveRecoverySchemes,
       }
       get(*lorentz_factor)[s] = primitive_data.value().lorentz_factor;
       get(*pressure)[s] = primitive_data.value().pressure;
+
+      get(*specific_enthalpy)[s] = primitive_data.value().rho_h_w_squared /
+                                   (primitive_data.value().lorentz_factor *
+                                    primitive_data.value().rest_mass_density);
+      get(*specific_internal_energy)[s] =
+          get(*specific_enthalpy)[s] - 1. -
```

Comment:
Floating point wise, it might be better to do $\rho h - \rho - p$ but :shrug: Both seem like they'll go poorly in the low-density limit. The only advantage is not dividing by $\rho$, something we've talked about trying to eliminate since in general it seems like $\rho h$ is what shows up, not $h$ as much

Reply:
I multiplied by rho.

One way might be to directly extract eps and T from the inversion, i.e., include in them in primitive_data where they will be computed most accurately.

Reply:
Ah yea, that's a good idea! For another PR, but that should help a lot. I've been worried about lower-density limits because of this sort of math in places.